pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskGenerator.java

```diff
@@ -0,0 +1,298 @@
+        compactionConfigs.getOrDefault(UpsertCompactionTask.INVALID_RECORDS_THRESHOLD_PERCENT,
+            String.valueOf(DEFAULT_INVALID_RECORDS_THRESHOLD_PERCENT)));
+    long invalidRecordsThresholdCount = Long.parseLong(
+        compactionConfigs.getOrDefault(UpsertCompactionTask.INVALID_RECORDS_THRESHOLD_COUNT,
+            String.valueOf(DEFAULT_INVALID_RECORDS_THRESHOLD_COUNT)));
+    List<SegmentZKMetadata> segmentsForCompaction = new ArrayList<>();
+    List<String> segmentsForDeletion = new ArrayList<>();
+    for (Map.Entry<String, String> streamResponse : responseSet) {
+      JsonNode allValidDocIdMetadata;
+      try {
+        allValidDocIdMetadata = JsonUtils.stringToJsonNode(streamResponse.getValue());
+      } catch (IOException e) {
+        LOGGER.error("Unable to parse validDocIdMetadata response for: {}", streamResponse.getKey());
+        continue;
+      }
+      Iterator<JsonNode> iterator = allValidDocIdMetadata.elements();
+      while (iterator.hasNext()) {
+        JsonNode validDocIdMetadata = iterator.next();
+        long invalidRecordCount = validDocIdMetadata.get("totalInvalidDocs").asLong();
+        String segmentName = validDocIdMetadata.get("segmentName").asText();
+        SegmentZKMetadata segment = completedSegmentsMap.get(segmentName);
+        double invalidRecordPercent = ((double) invalidRecordCount / segment.getTotalDocs()) * 100;
+        if (invalidRecordCount == segment.getTotalDocs()) {
+          segmentsForDeletion.add(segment.getSegmentName());
+        } else if (invalidRecordPercent > invalidRecordsThresholdPercent
+            && invalidRecordCount > invalidRecordsThresholdCount) {
+          segmentsForCompaction.add(segment);
+        }
+      }
+    }
+    return new SegmentSelectionResult(segmentsForCompaction, segmentsForDeletion);
+  }
+
+  @VisibleForTesting
+  public static List<String> getValidDocIdMetadataUrls(Map<String, List<String>> serverToSegments,
+      BiMap<String, String> serverToEndpoints, String tableNameWithType, Set<String> completedSegments)
+      throws URISyntaxException {
+    Set<String> remainingSegments = new HashSet<>(completedSegments);
+    List<String> urls = new ArrayList<>();
+    for (Map.Entry<String, List<String>> entry : serverToSegments.entrySet()) {
+      if (remainingSegments.isEmpty()) {
+        break;
+      }
+      String server = entry.getKey();
+      List<String> segmentNames = entry.getValue();
+      URIBuilder uriBuilder = new URIBuilder(serverToEndpoints.get(server)).setPath(
+          String.format("/tables/%s/validDocIdMetadata", tableNameWithType));
+      int completedSegmentCountPerServer = 0;
+      for (String segmentName : segmentNames) {
+        if (remainingSegments.contains(segmentName)) {
+          completedSegmentCountPerServer++;
+          uriBuilder.addParameter("segmentNames", segmentName);
+          remainingSegments.remove(segmentName);
+        }
```

Comment:
(minor) Can be simplified
```suggestion
        if (remainingSegments.remove(segmentName)) {
          completedSegmentCountPerServer++;
          uriBuilder.addParameter("segmentNames", segmentName);
        }
```